src/frontend/src/optimizer/mod.rs

```diff
@@ -294,8 +295,35 @@ impl PlanRoot {
             Convention::Logical => {
                 let plan = self.gen_optimized_logical_plan_for_stream()?;
 
-                let (plan, out_col_change) =
-                    plan.logical_rewrite_for_stream(&mut Default::default())?;
+                let (plan, out_col_change) = {
+                    let (plan, out_col_change) =
+                        plan.logical_rewrite_for_stream(&mut Default::default())?;
+                    if out_col_change.is_injective() {
+                        (plan, out_col_change)
+                    } else {
+                        let mut output_indices = (0..plan.schema().len()).collect_vec();
+                        #[allow(unused_assignments)]
+                        let (mut map, mut target_size) = out_col_change.into_parts();
+
+                        // TODO(st1page): https://github.com/risingwavelabs/risingwave/issues/7234
+                        // assert_eq!(target_size, output_indices.len());
+                        target_size = plan.schema().len();
+                        let mut tar_exists = vec![false; target_size];
+                        for i in map.iter_mut().flatten() {
```

Comment:
I'm not so sure how robust this method is. I don't think we can ensure the ordering of the columns in this case.

An actual solution is to ignore a line below:

```rust
self.out_fields = out_col_change.rewrite_bitset(&self.out_fields);
```

But I'm not so comfortable either. I prefer instead to fix problem at the root (i.e. similar to https://github.com/risingwavelabs/risingwave/pull/8267). For this PR, perhaps we can introduce an `debug_assert` injectivity check to the `out_col_change` here.

Reply:
For that reason alone, I think it's safer for us to avoid non-injectivity, unless we can demonstrate that mapping to an alias will never result in a broken chain of assumptions.

(but this entire issue originates from the fact that it does).

Reply:
Somehow, I get the feeling that this optimization should not happen in the `logical_rewrite_for_stream` step, but rather, as an optimization pass.

Reply:
I know your concern here. But the non-injection mapping here can not bring a question. The issue can only be with the non-surjection. Let's think about the changes_column here.
non-injection mapping: `[0->1, 1->1]` your original cloumn[0] and column[1] are at position 1 now, :heavy_check_mark: 
non-surjection mapping: `[0->0, 0->1]` your original cloumn[0] could be position 0 or 1, you can choose one. :x:

Reply:
Created a proposal here: https://github.com/risingwavelabs/risingwave/issues/8277